hazelcast-client-new/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -89,22 +113,68 @@
 
     private final AtomicInteger completionIdCounter = new AtomicInteger();
 
-    protected AbstractClientInternalCacheProxy(CacheConfig cacheConfig, ClientContext clientContext) {
+    protected AbstractClientInternalCacheProxy(CacheConfig cacheConfig, ClientContext clientContext,
+                                               HazelcastClientCacheManager cacheManager) {
         super(cacheConfig, clientContext);
+        this.cacheManager = cacheManager;
+        nearCacheManager = clientContext.getNearCacheManager();
         asyncListenerRegistrations = new ConcurrentHashMap<CacheEntryListenerConfiguration, String>();
         syncListenerRegistrations = new ConcurrentHashMap<CacheEntryListenerConfiguration, String>();
         syncLocks = new ConcurrentHashMap<Integer, CountDownLatch>();
 
         NearCacheConfig nearCacheConfig = cacheConfig.getNearCacheConfig();
         if (nearCacheConfig != null) {
-            nearCache = new ClientHeapNearCache<Data>(nameWithPrefix, clientContext, nearCacheConfig);
             cacheOnUpdate = nearCacheConfig.getLocalUpdatePolicy() == NearCacheConfig.LocalUpdatePolicy.CACHE;
+            NearCacheContext nearCacheContext =
+                    new NearCacheContext(clientContext.getSerializationService(),
+                            createNearCacheExecutor(clientContext.getExecutionService()));
+            nearCache = nearCacheManager
+                    .getOrCreateNearCache(nameWithPrefix, nearCacheConfig, nearCacheContext);
+            registerInvalidationListener();
         } else {
             nearCache = null;
             cacheOnUpdate = false;
         }
     }
 
+    protected class ClientNearCacheExecutor implements Executor {
+
+        protected ClientExecutionService clientExecutionService;
+
+        protected ClientNearCacheExecutor(ClientExecutionService clientExecutionService) {
+            this.clientExecutionService = clientExecutionService;
+        }
+
+        @Override
+        public void execute(Runnable runnable) {
+            clientExecutionService.execute(runnable);
+        }
+
+    }
+
+    protected Executor createNearCacheExecutor(ClientExecutionService clientExecutionService) {
+        return new ClientNearCacheExecutor(clientExecutionService);
+    }
+
+    @Override
+    public void close() {
+        if (nearCache != null) {
+            removeInvalidationListener();
+            nearCacheManager.clearNearCache(nearCache.getName());
```

Comment:
currently, there is no specific stuff on ```NearCacheManager``` for clearing cache. But I prefer giving responsibility for such kind of operations like clear or destroy to manager